src/caishen_dashboard/data_processing/technical_indicators.py

```diff
@@ -0,0 +1,197 @@
+    lookback = 20 calculates SMA20
+
+    Args:
+        values (List[float]): The list of float values to average
+        lookback (int, optional): The lookback. Defaults to 14.
+
+    Raises:
+        InvalidInputError: The list "values" must not be empty
+        InvalidInputError: The lookback has to be an int
+        InvalidInputError: The lookback has to be non negative
+
+    Returns:
+        List[float]: The list of SMA values. This list has the same length as "values"
+                     The first "lookback - 1" elements are set to -1 to indicate there
+                     is no real calculation to be done for them
+    Example:
+        >>> from caishen_dashboard.data_processing.technical_indicators import SMA
+        >>> SMA([1.0,2.0,3.0,4.0],3)
+        [-1, -1, 2.0, 3.0]
+    """
+
+
+    # Error Checking
+    if len(values) == 0:
+        raise InvalidInputError("The length of the values list is 0. It should be at least 1")
+    if type(lookback) is not int:
+        raise InvalidInputError("The lookback is expected to be an int, but it's type is " + str(type(lookback)))
+    if lookback < 0:
+        raise InvalidInputError("The lookback value has to be a non negative integer, but it is set to " + str(lookback))
+
+    result: List[float] = []
+
+    # calculate SMA for all values except the first lookback ones
+    for i in range(lookback - 1):
+        result.append(-1)
+    for i in range(len(values) - lookback + 1):
+        average = sum(values[i:i + lookback]) / (1.0*lookback)
+        result.append(average)
+
+    return result
+
+def EMA(values: List[float], lookback: int = 12, smoothing: float = 2.0):
+    """Calculates Exponential Moving Average (EMA) for a given lookback.
+    
+    The lookback determines the rolling window for which the averages are taken over. For example,
+    lookback = 20 calculates EMA20
+
+    The formula for EMA is EMA(t) = price(t) * multiplier + EMA(t-1)*(1-multiplier)
+    where multiplier = smoothing / (1 + MIN(# of entries,lookback))
+
+    Args:
+        values (List[float]): The list of float values to average
+        lookback (int, optional): The lookback. . Defaults to 12.
+        smoothing (float, optional): Smoothing factor. Defaults to 2.0.
+
+    Raises:
+        InvalidInputError: The list "values" must not be empty
+        InvalidInputError: The lookback has to be an int
+        InvalidInputError: The lookback has to be non negative
+        InvalidInputError: The smoothing has to be non negative
+
+    Returns:
+        List[float]: The list of EMA values. This list has the same length as "values"
+                     The first "lookback - 1" elements are set to -1 to indicate there
+                     is no real calculation to be done for them
+
+    Example:
+        >>> from caishen_dashboard.data_processing.technical_indicators import EMA
+        >>> EMA([1.0,2.0,3.0,4.0,5.0],3,2.0)
+        [-1, -1, 2.0, 3.0, 4.0]
```

Comment:
I believe the list should be ordered as [-1, -1, 1.5, 2.75, 3.875]
Multiplier is 0.5 (2/(1+3))
t =3 
EMA(t) = 3* 0.5 + 0  = 1.5
(Second value is 0 because there is no EMA(t-1))
t=4
EMA(t) = 4 * 0.5 + 1.5 * 0.5 = 2.75
...

Reply:
Might be better to give an example with n = 4 or smoothing  = 1 to avoid 0.5 multipliers for both values

Reply:
> I believe the list should be ordered as [-1, -1, 1.5, 2.75, 3.875]
> Multiplier is 0.5 (2/(1+3))
> t =3
> EMA(t) = 3* 0.5 + 0 = 1.5
> (Second value is 0 because there is no EMA(t-1))
> t=4
> EMA(t) = 4 * 0.5 + 1.5 * 0.5 = 2.75
> ...

Ema(lookback) = SMA(lookback) aka the average of the first lookback numbers, so this calculation is correct
https://www.educba.com/exponential-moving-average-formula/